src/world.js

```diff
@@ -231,6 +232,187 @@ $.extend( $.World.prototype, $.EventSource.prototype, /** @lends OpenSeadragon.W
+        // However, we must pick the correct drawer reference (navigator VS viewer)
+        const supportedFormats = this.viewer.drawer.getSupportedDataFormats();
+        const keepInternalCacheCopy = this.viewer.drawer.options.usePrivateCache;
+        const drawerId = this.viewer.drawer.getId();
+
+        const jobList = tileList.map(tile => {
+            const tiledImage = tile.tiledImage;
+            const originalCache = tile.getCache(tile.originalCacheKey);
+            let workingCache = null;
+            const getWorkingCacheData = (type) => {
+                if (workingCache) {
+                    return workingCache.getDataAs(type, false);
+                }
+
+                const targetCopyKey = restoreTiles ? tile.originalCacheKey : tile.cacheKey;
+                const origCache = tile.getCache(targetCopyKey);
+                if (!origCache) {
+                    $.console.error("[Tile::getData] There is no cache available for tile with key %s", targetCopyKey);
+                    return $.Promise.reject();
+                }
+                // Here ensure type is defined, rquired by data callbacks
+                type = type || origCache.type;
+                workingCache = new $.CacheRecord().withTileReference(tile);
+                return origCache.getDataAs(type, true).then(data => {
+                    workingCache.addTile(tile, data, type);
+                    return workingCache.data;
+                });
+            };
+            const setWorkingCacheData = (value, type) => {
+                if (!workingCache) {
+                    workingCache = new $.CacheRecord().withTileReference(tile);
+                    workingCache.addTile(tile, value, type);
+                } else {
+                    workingCache.setDataAs(value, type);
+                }
+            };
+            const atomicCacheSwap = () => {
+                if (workingCache) {
+                    let newCacheKey = tile.buildDistinctMainCacheKey();
+                    tiledImage._tileCache.injectCache({
+                        tile: tile,
+                        cache: workingCache,
+                        targetKey: newCacheKey,
+                        setAsMainCache: true,
+                        tileAllowNotLoaded: false //todo what if called from load event?
+                    });
+                } else if (restoreTiles) {
+                    // If we requested restore, perform now
+                    tiledImage._tileCache.restoreTilesThatShareOriginalCache(tile, tile.getCache(tile.originalCacheKey), true);
+                }
+            };
+
+            //todo docs
+            return eventTarget.raiseEventAwaiting('tile-invalidated', {
+                tile: tile,
+                tiledImage: tiledImage,
+                outdated: () => originalCache.__invStamp !== tStamp,
+                getData: getWorkingCacheData,
+                setData: setWorkingCacheData,
```

Comment:
I realized that I tried to push working cache onto the tile reference. But, each event should instead keep its own reference to avoid colliding and update accordingly. Now, the modification events for data are bound to the event itself, and support get/set for data, `outdated()` flag that is true when a newer version of the invalidation routine runs on the tile too, and `resetData()` that one can use to throw away any prior modifications within that event.

It also simplifies some parts of the logics, since we do not attach the working cache to the cache container (`tileCache`) until we are done processing (`injectCache()`). Now, some tileCache methods (`cloneCache` or `replaceCache`) are no longer used, but they might be useful... I did not really want to remove them. Current approach removes `setData` and `getData` from tile class, meaning tile only knows caches (no data-level abstraction), so if someone wants to do custom stuff outside the `tile-invalidated` routine, they have to manage cache manually, and cloning or replacing on the `tileCache` might be handy..

Reply:
The current filtering plugin logics shrank to:

````
const contextCopy = await e.getData('context2d');
if (!contextCopy) return;

for (let i = 0; i < processors.length; i++) {
    if (e.outdated()) return;
    await processors[i](contextCopy);
}
if (e.outdated()) return;
await e.setData(contextCopy, 'context2d');
````
note that we no longer use `tile` referece

Reply:
I can't say I understand all the implications, but this direction certainly sounds nice and clean.